File: codebase_scanner/test_vulnerable.js
// JavaScript test file with vulnerabilities

const express = require('express');
const app = express();

// XSS vulnerability
app.get('/search', (req, res) => {
    const query = req.query.q;
    // VULNERABLE: Direct HTML rendering without escaping
    res.send(`<h1>Search results for: ${query}</h1>`);
});

// SQL Injection 
app.get('/user/:id', (req, res) => {
    const userId = req.params.id;
    // VULNERABLE: Direct string concatenation
    db.query(`SELECT * FROM users WHERE id = ${userId}`, (err, results) => {
        res.json(results);
    });
});

// Hardcoded secrets
const API_SECRET = "super-secret-key-123";
const JWT_SECRET = "jwt-secret-key";

// Path traversal
app.get('/file', (req, res) => {
    const filename = req.query.name;
    // VULNERABLE: No path validation
    res.sendFile(`/uploads/${filename}`);
});

// Insecure randomness
function generateToken() {
    // VULNERABLE: Using Math.random() for security
    return Math.random().toString(36).substring(7);
}

File: codebase_scanner/test_vulnerable.py
import os
import sqlite3
import pickle

# SQL Injection vulnerability
def get_user(user_id):
    conn = sqlite3.connect('users.db')
    query = f"SELECT * FROM users WHERE id = {user_id}"  # SQL Injection
    return conn.execute(query).fetchall()

# Hardcoded password
DB_PASSWORD = "admin123"  # Hardcoded credential
API_KEY = "sk-1234567890abcdef"  # Hardcoded API key

# Command injection
def run_command(user_input):
    os.system(f"echo {user_input}")  # Command injection

# Insecure deserialization
def load_data(data):
    return pickle.loads(data)  # CWE-502: Insecure deserialization